// @geometricpanda/storybook-addon-badges/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var r="@geometricpanda/storybook-addon-badges",l=r,d="Storybook Addon Badges",s="badgesConfig",E="badges",e=(o=>(o.DEFAULT="default",o.BETA="beta",o.STABLE="stable",o.NEEDS_REVISION="needs-revision",o.OBSOLETE="obsolete",o.EXPERIMENTAL="experimental",o.DEPRECATED="deprecated",o))(e||{});var t={["default"]:{title:"Badge"},["beta"]:{title:"Beta",styles:{backgroundColor:"#D6E0FF",borderColor:"#2952CC",color:"#2952CC"}},["stable"]:{title:"Stable",styles:{backgroundColor:"#DCF2EA",borderColor:"#317159",color:"#317159"}},["needs-revision"]:{title:"Needs Revision",styles:{backgroundColor:"#FFEFD2",borderColor:"#66460D",color:"#66460D"}},["obsolete"]:{title:"Obsolete",styles:{backgroundColor:"#F9DADA",borderColor:"#7D2828",color:"#7D2828"}},["experimental"]:{title:"Experimental",styles:{backgroundColor:"#E7E4F9",borderColor:"#6E62B6",color:"#6E62B6"}},["deprecated"]:{title:"Deprecated",styles:{backgroundColor:"#F8E3DA",borderColor:"#85462B",color:"#85462B"}}},g=t["default"];var i={};

exports.ADDON_ID = r;
exports.ADDON_TITLE = d;
exports.BADGE = e;
exports.PARAM_BADGES_KEY = E;
exports.PARAM_CONFIG_KEY = s;
exports.TOOL_ID = l;
exports.default = i;
exports.defaultBadgeConfig = g;
exports.defaultBadgesConfig = t;
