// @nx/storybook/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initGeneratorInternal = exports.initGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const update_package_scripts_1 = require("@nx/devkit/src/utils/update-package-scripts");
const semver_1 = require("semver");
const plugin_1 = require("../../plugins/plugin");
const utilities_1 = require("../../utils/utilities");
const versions_1 = require("../../utils/versions");
const update_gitignore_1 = require("./lib/update-gitignore");
function checkDependenciesInstalled(host, schema) {
    const devDependencies = {
        '@nx/storybook': versions_1.nxVersion,
        '@nx/web': versions_1.nxVersion,
    };
    if (schema.addPlugin) {
        let storybook7VersionToInstall = versions_1.storybookVersion;
        if ((0, utilities_1.storybookMajorVersion)() >= 7 &&
            (0, utilities_1.getInstalledStorybookVersion)() &&
            (0, semver_1.gte)((0, utilities_1.getInstalledStorybookVersion)(), '7.0.0')) {
            storybook7VersionToInstall = (0, utilities_1.getInstalledStorybookVersion)();
        }
        devDependencies['storybook'] = storybook7VersionToInstall;
    }
    return (0, devkit_1.addDependenciesToPackageJson)(host, {}, devDependencies, undefined, schema.keepExistingVersions);
}
function addCacheableOperation(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    const cacheableOperations = nxJson.tasksRunnerOptions?.default?.options?.cacheableOperations;
    if (cacheableOperations && cacheableOperations.includes('build-storybook')) {
        nxJson.tasksRunnerOptions.default.options.cacheableOperations.push('build-storybook');
    }
    nxJson.targetDefaults ??= {};
    nxJson.targetDefaults['build-storybook'] ??= {};
    nxJson.targetDefaults['build-storybook'].cache = true;
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function moveToDevDependencies(tree) {
    let updated = false;
    (0, devkit_1.updateJson)(tree, 'package.json', (packageJson) => {
        packageJson.dependencies = packageJson.dependencies || {};
        packageJson.devDependencies = packageJson.devDependencies || {};
        if (packageJson.dependencies['@nx/storybook']) {
            packageJson.devDependencies['@nx/storybook'] =
                packageJson.dependencies['@nx/storybook'];
            delete packageJson.dependencies['@nx/storybook'];
            updated = true;
        }
        return packageJson;
    });
    return updated ? () => (0, devkit_1.installPackagesTask)(tree) : () => { };
}
function initGenerator(tree, schema) {
    return initGeneratorInternal(tree, { addPlugin: false, ...schema });
}
exports.initGenerator = initGenerator;
async function initGeneratorInternal(tree, schema) {
    schema.addPlugin ??= process.env.NX_ADD_PLUGINS !== 'false';
    if (schema.addPlugin) {
        (0, utilities_1.addPlugin)(tree);
        (0, update_gitignore_1.updateGitignore)(tree);
    }
    else {
        addCacheableOperation(tree);
    }
    const tasks = [];
    if (!schema.skipPackageJson) {
        tasks.push(moveToDevDependencies(tree));
        tasks.push(checkDependenciesInstalled(tree, schema));
    }
    if (schema.updatePackageScripts) {
        await (0, update_package_scripts_1.updatePackageScripts)(tree, plugin_1.createNodes);
    }
    if (!schema.skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
exports.initGeneratorInternal = initGeneratorInternal;
exports.default = initGenerator;
