// @nxext/stencil/src/utils/lint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStencilEslintJson = exports.extraEslintDependencies = void 0;
const versions_1 = require("./versions");
const devkit_1 = require("@nx/devkit");
exports.extraEslintDependencies = {
    dependencies: {},
    devDependencies: {
        'eslint-plugin-import': versions_1.eslintPluginImportVersion,
        '@stencil/eslint-plugin': versions_1.eslintPluginStencilVersion,
    },
};
const createStencilEslintJson = (projectRoot) => ({
    extends: [
        'plugin:@stencil/recommended',
        'plugin:import/recommended',
        'plugin:import/typescript',
        `${(0, devkit_1.offsetFromRoot)(projectRoot)}.eslintrc.json`,
    ],
    ignorePatterns: ['!**/*'],
    overrides: [
        {
            files: ['*.ts', '*.tsx'],
            /**
             * Having an empty rules object present makes it more obvious to the user where they would
             * extend things from if they needed to
             */
            rules: {},
        },
        {
            files: ['*.ts', '*.tsx'],
            rules: {},
        },
        {
            files: ['*.js', '*.jsx'],
            rules: {},
        },
    ],
});
exports.createStencilEslintJson = createStencilEslintJson;
