// @nxext/stencil/src/utils/typings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STYLE_PLUGIN_DEPENDENCIES = exports.PROJECT_TYPE_DEPENDENCIES = exports.AppType = void 0;
const versions_1 = require("./versions");
const stencil_core_utils_1 = require("../stencil-core-utils");
var AppType;
(function (AppType) {
    AppType["application"] = "application";
    AppType["library"] = "library";
    AppType["pwa"] = "pwa";
})(AppType || (exports.AppType = AppType = {}));
const stencilDependencies = {
    '@stencil/core': versions_1.stencilVersion,
};
const e2ePuppeteerDependencies = {
    puppeteer: versions_1.puppeteer,
};
exports.PROJECT_TYPE_DEPENDENCIES = {
    [AppType.application]: {
        dependencies: {},
        devDependencies: Object.assign(Object.assign({}, stencilDependencies), { '@stencil/router': versions_1.stencilRouterVersion }),
    },
    [AppType.pwa]: {
        dependencies: {},
        devDependencies: Object.assign(Object.assign({}, stencilDependencies), { '@ionic/core': versions_1.ionicVersion }),
    },
    [AppType.library]: {
        dependencies: {},
        devDependencies: Object.assign({}, stencilDependencies),
    },
    init: {
        dependencies: {},
        devDependencies: Object.assign({}, stencilDependencies),
    },
    puppeteer: {
        dependencies: {},
        devDependencies: Object.assign({}, e2ePuppeteerDependencies),
    },
};
exports.STYLE_PLUGIN_DEPENDENCIES = {
    [stencil_core_utils_1.SupportedStyles.css]: {
        dependencies: {},
        devDependencies: {},
    },
    [stencil_core_utils_1.SupportedStyles.less]: {
        dependencies: {},
        devDependencies: {
            '@stencil/less': versions_1.STENCIL_STYLE_PLUGIN_VERSION[stencil_core_utils_1.SupportedStyles.less],
        },
    },
    [stencil_core_utils_1.SupportedStyles.scss]: {
        dependencies: {},
        devDependencies: {
            '@stencil/sass': versions_1.STENCIL_STYLE_PLUGIN_VERSION[stencil_core_utils_1.SupportedStyles.scss],
        },
    },
    [stencil_core_utils_1.SupportedStyles.styl]: {
        dependencies: {},
        devDependencies: {
            '@stencil/stylus': versions_1.STENCIL_STYLE_PLUGIN_VERSION[stencil_core_utils_1.SupportedStyles.styl],
        },
    },
    [stencil_core_utils_1.SupportedStyles.pcss]: {
        dependencies: {},
        devDependencies: {
            '@stencil/postcss': versions_1.STENCIL_STYLE_PLUGIN_VERSION[stencil_core_utils_1.SupportedStyles.pcss],
            autoprefixer: versions_1.autoprefixerVersion,
            '@types/autoprefixer': versions_1.autoprefixerTypesVersion,
        },
    },
};
