// @nxext/stencil/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nxextSvelteVersion = exports.STENCIL_STYLE_PLUGIN_VERSION = exports.STENCIL_OUTPUTTARGET_VERSION = exports.eslintPluginStencilVersion = exports.eslintPluginImportVersion = exports.puppeteer = exports.autoprefixerTypesVersion = exports.autoprefixerVersion = exports.ionicVersion = exports.stencilRouterVersion = exports.stencilVersion = void 0;
exports.stencilVersion = '^4.12.6';
exports.stencilRouterVersion = '^1.0.1';
exports.ionicVersion = '^6.4.1';
exports.autoprefixerVersion = '^9.7.6';
exports.autoprefixerTypesVersion = '^9.7.2';
exports.puppeteer = '^21.4.1';
exports.eslintPluginImportVersion = '2.22.1';
exports.eslintPluginStencilVersion = '0.3.1';
exports.STENCIL_OUTPUTTARGET_VERSION = {
    react: '^0.5.3',
    angular: '^0.8.3',
    vue: '^0.8.6',
    svelte: '^0.0.3',
};
exports.STENCIL_STYLE_PLUGIN_VERSION = {
    less: '1.0.0',
    scss: '3.0.7',
    pcss: '2.1.0',
    styl: '1.0.1',
};
exports.nxextSvelteVersion = '^16.4.0';
