// @telenet/wink-core/dist/collection/components/input/input.js
import { Fragment, Host, h } from "@stencil/core";
import { INPUT_APPEARANCES, INPUT_AUTOCOMPLETES, INPUT_ENTERKEYHINTS, INPUT_MODES, INPUT_SIZES, INPUT_TYPES, } from "./models/input.types";
import { ICON_SIZES } from "../icon/models/icon.types";
export class Input {
    constructor() {
        this.appearance = INPUT_APPEARANCES.default;
        this.value = '';
        this.inputId = undefined;
        this.name = undefined;
        this.type = INPUT_TYPES.text;
        this.placeholder = '';
        this.isRequired = false;
        this.inputModeOption = INPUT_MODES.text;
        this.autocomplete = INPUT_AUTOCOMPLETES.off;
        this.enterkeyhintOption = INPUT_ENTERKEYHINTS.none;
        this.isReadonly = false;
        this.isDisabled = false;
        this.isAutofocused = false;
        this.error = '';
        this.prefixValue = '';
        this.postfixValue = '';
        this.iconLeft = '';
        this.iconRight = '';
        this.showIconRight = true;
        this.inputWidth = 0;
        this.label = '';
        this.helper = '';
        this.size = INPUT_SIZES.md;
    }
    getLabelId() {
        return this.inputId + ':Label';
    }
    getHelperId() {
        return this.inputId + ':Helper';
    }
    getErrorMsgId() {
        return this.inputId + ':Error';
    }
    getAreaDescribedby() {
        return [
            this.label && this.getLabelId(),
            this.helper && this.getHelperId(),
            this.error?.length > 0 && this.getErrorMsgId(),
        ]
            .filter((item) => typeof item === 'string' && item.trim().length > 0)
            .join(' ');
    }
    get getLabel() {
        if (this.label) {
            return (h(Fragment, null, h("label", { htmlFor: this.inputId, id: this.getLabelId(), class: { label: true, 'label-disabled': this.isDisabled } }, h("div", { class: "label-label" }, this.label, this.helper && (h("div", { class: "label-helper", id: this.getHelperId() }, this.helper))), !this.isRequired && h("div", { class: "label-optional" }, "Optional"))));
        }
    }
    get getPrefix() {
        if (this.prefixValue && !this.iconLeft) {
            return (h(Fragment, null, h("div", { class: {
                    'input-affix input-affix-prefix': true,
                    'input-affix-disabled': this.isDisabled,
                    'input-affix-readonly': this.isReadonly,
                } }, this.prefixValue)));
        }
    }
    get getPostfix() {
        if (this.postfixValue && !this.iconRight) {
            return (h(Fragment, null, h("div", { class: {
                    'input-affix input-affix-postfix': true,
                    'input-affix-disabled': this.isDisabled,
                    'input-affix-readonly': this.isReadonly,
                } }, this.postfixValue)));
        }
    }
    getReadOnlyLength(value) {
        if (this.isReadonly && value.length > 0) {
            this.inputWidth = value.length;
        }
        else {
            this.inputWidth = 0;
        }
    }
    showSearchBackSpaceIcon(value) {
        if (value.length === 0 && this.appearance === INPUT_APPEARANCES.search) {
            this.showIconRight = false;
        }
        else {
            this.showIconRight = true;
        }
    }
    handleChange(event) {
        this.value = event.target.value;
        this.showSearchBackSpaceIcon(this.value);
        this.getReadOnlyLength(this.value);
    }
    componentDidRender() {
        this.showSearchBackSpaceIcon(this.value);
        this.getReadOnlyLength(this.value);
    }
    render() {
        return (h(Host, { key: '418047881e8349b8ae179ee60710e226241742d6' }, this.getLabel, h("div", { key: 'a71fde80ade31601ec2d740d9ea0d19b8d8449e0', class: {
                'input-wrapper': true,
                'input-with-icon-button input-with-icon-button-left': this.iconLeft?.length > 0,
                'input-with-icon-button input-with-icon-button-right': this.iconRight?.length > 0,
            } }, this.getPrefix, this.iconLeft?.length > 0 && (h("div", { class: "input-icon-button input-icon-button-left" }, h("button", { type: "button", "aria-label": this.iconLeft, class: "input-icon-button-button" }, h("wink-icon", { appearance: "default", icon: this.iconLeft, size: this.size })))), h("input", { key: '241c76b85aa39a1552b1cb58a75429342a9960d9', onInput: (e) => this.handleChange(e), id: this.inputId, name: this.name, type: this.type, value: this.value, placeholder: this.placeholder, "input-mode": this.inputModeOption, autocomplete: this.autocomplete, autofocus: this.isAutofocused, enterkeyhint: this.enterkeyhintOption, required: this.isRequired, readonly: this.isReadonly, disabled: this.isDisabled, "aria-describedby": this.getAreaDescribedby(), "aria-invalid": this.error?.length > 0, class: {
                input: true,
            }, style: {
                width: this.isReadonly && this.inputWidth > 0 ? `${this.inputWidth}ch` : '',
            } }), this.iconRight?.length > 0 && this.showIconRight && (h("div", { class: "input-icon-button input-icon-button-right" }, h("button", { type: "button", "aria-label": this.iconRight, class: "input-icon-button-button" }, h("wink-icon", { appearance: "default", icon: this.iconRight, size: this.size })))), this.getPostfix), this.error?.length > 0 && (h("div", { id: this.getErrorMsgId(), class: "error-msg" }, h("wink-icon", { appearance: "default", icon: "error-shape", size: ICON_SIZES.md }), h("div", { class: "error-label font-caption " }, this.error)))));
    }
    static get is() { return "wink-input"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "telenet": ["styles/brands/input.telenet.scss"],
            "base": ["styles/brands/input.base.scss"]
        };
    }
    static get styleUrls() {
        return {
            "telenet": ["styles/brands/input.telenet.css"],
            "base": ["styles/brands/input.base.css"]
        };
    }
    static get properties() {
        return {
            "appearance": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "InputAppearance",
                    "resolved": "\"default\" | \"search\"",
                    "references": {
                        "InputAppearance": {
                            "location": "import",
                            "path": "./models/input.types",
                            "id": "../core/src/components/input/models/input.types.tsx::InputAppearance"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "appearance",
                "reflect": true,
                "defaultValue": "INPUT_APPEARANCES.default"
            },
            "value": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "value",
                "reflect": true,
                "defaultValue": "''"
            },
            "inputId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "input-id",
                "reflect": false
            },
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "name",
                "reflect": false
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "InputType",
                    "resolved": "\"date\" | \"number\" | \"password\" | \"tel\" | \"text\" | \"time\" | \"url\"",
                    "references": {
                        "InputType": {
                            "location": "import",
                            "path": "./models/input.types",
                            "id": "../core/src/components/input/models/input.types.tsx::InputType"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "type",
                "reflect": false,
                "defaultValue": "INPUT_TYPES.text"
            },
            "placeholder": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "placeholder",
                "reflect": false,
                "defaultValue": "''"
            },
            "isRequired": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "is-required",
                "reflect": false,
                "defaultValue": "false"
            },
            "inputModeOption": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "InputMode",
                    "resolved": "\"decimal\" | \"email\" | \"none\" | \"numeric\" | \"search\" | \"tel\" | \"text\" | \"url\"",
                    "references": {
                        "InputMode": {
                            "location": "import",
                            "path": "./models/input.types",
                            "id": "../core/src/components/input/models/input.types.tsx::InputMode"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "input-mode-option",
                "reflect": false,
                "defaultValue": "INPUT_MODES.text"
            },
            "autocomplete": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "InputAutocomplete",
                    "resolved": "\"address\" | \"current-password\" | \"email\" | \"name\" | \"new-password\" | \"off\" | \"on\" | \"username\"",
                    "references": {
                        "InputAutocomplete": {
                            "location": "import",
                            "path": "./models/input.types",
                            "id": "../core/src/components/input/models/input.types.tsx::InputAutocomplete"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "autocomplete",
                "reflect": false,
                "defaultValue": "INPUT_AUTOCOMPLETES.off"
            },
            "enterkeyhintOption": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "InputEnterkeyhint",
                    "resolved": "\"done\" | \"enter\" | \"go\" | \"next\" | \"none\" | \"previous\" | \"search\" | \"send\"",
                    "references": {
                        "InputEnterkeyhint": {
                            "location": "import",
                            "path": "./models/input.types",
                            "id": "../core/src/components/input/models/input.types.tsx::InputEnterkeyhint"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "enterkeyhint-option",
                "reflect": false,
                "defaultValue": "INPUT_ENTERKEYHINTS.none"
            },
            "isReadonly": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "is-readonly",
                "reflect": false,
                "defaultValue": "false"
            },
            "isDisabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "is-disabled",
                "reflect": false,
                "defaultValue": "false"
            },
            "isAutofocused": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "is-autofocused",
                "reflect": false,
                "defaultValue": "false"
            },
            "error": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "error",
                "reflect": false,
                "defaultValue": "''"
            },
            "prefixValue": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "prefix-value",
                "reflect": false,
                "defaultValue": "''"
            },
            "postfixValue": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "postfix-value",
                "reflect": false,
                "defaultValue": "''"
            },
            "iconLeft": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "icon-left",
                "reflect": false,
                "defaultValue": "''"
            },
            "iconRight": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "icon-right",
                "reflect": false,
                "defaultValue": "''"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "label",
                "reflect": false,
                "defaultValue": "''"
            },
            "helper": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "helper",
                "reflect": false,
                "defaultValue": "''"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "InputSize",
                    "resolved": "string",
                    "references": {
                        "InputSize": {
                            "location": "import",
                            "path": "./models/input.types",
                            "id": "../core/src/components/input/models/input.types.tsx::InputSize"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "size",
                "reflect": true,
                "defaultValue": "INPUT_SIZES.md"
            }
        };
    }
    static get states() {
        return {
            "showIconRight": {},
            "inputWidth": {}
        };
    }
}
