// @nxext/stencil/src/executors/stencil-runtime/e2e-testing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cleanupE2eTesting = exports.prepareE2eTesting = void 0;
const fileutils_1 = require("../../utils/fileutils");
const fileutils_2 = require("@nx/workspace/src/utilities/fileutils");
const path_1 = require("path");
const devkit_1 = require("@nx/devkit");
function prepareE2eTesting(pathCollection) {
    const libPackageJson = {
        name: pathCollection.projectName,
        version: '1',
        main: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/index.cjs.js`,
        module: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/index.js`,
        es2015: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/esm/index.mjs`,
        es2017: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/esm/index.mjs`,
        types: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/types/index.d.ts`,
        collection: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/collection/collection-manifest.json`,
        'collection:main': `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/collection/index.js`,
        unpkg: `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/${pathCollection.projectName}/${pathCollection.projectName}.js`,
        files: [
            `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/dist/`,
            `${(0, path_1.relative)(pathCollection.projectRoot, pathCollection.distDir)}/loader/`,
        ],
    };
    (0, devkit_1.writeJsonFile)((0, devkit_1.joinPathFragments)(`${pathCollection.projectRoot}/package.e2e.json`), libPackageJson);
}
exports.prepareE2eTesting = prepareE2eTesting;
function cleanupE2eTesting(pathCollection) {
    const pkgJsonPath = (0, devkit_1.joinPathFragments)(`${pathCollection.projectRoot}/package.e2e.json`);
    if ((0, fileutils_2.fileExists)(pkgJsonPath)) {
        (0, fileutils_1.deleteFile)(pkgJsonPath);
    }
}
exports.cleanupE2eTesting = cleanupE2eTesting;
