// @nx/storybook/src/migrations/update-15-7-0/add-addon-essentials-to-all.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRootMainVariableName = void 0;
const devkit_1 = require("@nx/devkit");
const executor_options_utils_1 = require("@nx/devkit/src/generators/executor-options-utils");
const tsquery_1 = require("@phenomnomnominal/tsquery");
const remove_root_config_1 = require("./remove-root-config");
/**
 * The purpose of this migrator is to help users move away
 * from the root .storybook/ configuration folder and files.
 *
 * Since we cannot be sure of how users make use of the root Storybook
 * directory, what we are going to do is the following:
 *
 * 1. Make sure that all project-level .storybook/main.js files contain
 * the @storybook/addon-essentials addon.
 * 2. If the root .storybook/main.js file contains the @storybook/addon-essentials remove it
 * from the root
 * 3. If there are things beyond the addons array in the root .storybook/main.js file,
 * then keep it as it is - inform user that they need to manually need to copy over any extra stuff
 * 4. If the root .storybook/main.js file is now empty, inform user that they can delete it safely
 *
 * Point the user to a guide that explains how to all these things.
 */
async function default_1(tree) {
    const projectsThatFailedTOAddAddonEssentials = addAddonEssentialsToAllStorybooks(tree);
    if (projectsThatFailedTOAddAddonEssentials.length) {
        devkit_1.logger.info(`
          We could not add the @storybook/addon-essentials addon
          to the following projects' Storybook configurations:
    
          ${projectsThatFailedTOAddAddonEssentials.join(', ')}
    
          Please add it manually in the addons array of your project's
          .storybook/main.js|ts file.
          `);
    }
    const rootMainJsTsPath = tree.exists('.storybook/main.js')
        ? '.storybook/main.js'
        : tree.exists('.storybook/main.ts')
            ? '.storybook/main.ts'
            : undefined;
    if (rootMainJsTsPath) {
        const addonArrayOrEssentialsRemoved = removeAddonEssentialsFromRootStorybook(tree, rootMainJsTsPath);
        const storiesArrayRemoved = removeStoriesArrayFromRootIfEmpty(tree, rootMainJsTsPath);
        const removedRoot = (0, remove_root_config_1.removeRootConfig)(tree, rootMainJsTsPath);
        if (removedRoot) {
            // Logs are already printed in the removeRootConfig function
        }
        else {
            devkit_1.logger.info(`
      We removed the ${addonArrayOrEssentialsRemoved === 'addons'
                ? 'addons array '
                : '@storybook/addon-essentials addon '} 
      ${storiesArrayRemoved ? 'and the stories array ' : ''}
      from the root .storybook/main.js|ts file.
      `);
        }
    }
    devkit_1.logger.info(`
    Read more about our effort to deprecate the root .storybook folder here:
    https://nx.dev/packages/storybook/documents/configuring-storybook
    `);
    await (0, devkit_1.formatFiles)(tree);
}
exports.default = default_1;
function removeAddonEssentialsFromRootStorybook(tree, rootMainJsTsPath) {
    let rootMainJsTs = tree.read(rootMainJsTsPath, 'utf-8');
    const addonEssentials = tsquery_1.tsquery.query(rootMainJsTs, 'StringLiteral:has([text="@storybook/addon-essentials"])')?.[0];
    if (addonEssentials?.getText()?.length) {
        const fullAddonsNode = tsquery_1.tsquery.query(rootMainJsTs, 'PropertyAssignment:has([name="addons"])')?.[0];
        const stringLiterals = tsquery_1.tsquery.query(fullAddonsNode, 'StringLiteral');
        if (stringLiterals?.length === 1 &&
            stringLiterals?.[0]?.getText() === `'@storybook/addon-essentials'`) {
            rootMainJsTs = (0, devkit_1.applyChangesToString)(rootMainJsTs, [
                {
                    type: devkit_1.ChangeType.Delete,
                    start: fullAddonsNode.getStart(),
                    length: rootMainJsTs[fullAddonsNode.getEnd()] === ','
                        ? fullAddonsNode.getText().length + 1
                        : fullAddonsNode.getText().length,
                },
            ]);
            tree.write(rootMainJsTsPath, rootMainJsTs);
            return 'addons';
        }
        else {
            rootMainJsTs = (0, devkit_1.applyChangesToString)(rootMainJsTs, [
                {
                    type: devkit_1.ChangeType.Delete,
                    start: addonEssentials.getStart(),
                    length: rootMainJsTs[addonEssentials.getEnd()] === ','
                        ? addonEssentials.getText().length + 1
                        : addonEssentials.getText().length,
                },
            ]);
            tree.write(rootMainJsTsPath, rootMainJsTs);
            return 'esssentials';
        }
    }
}
function addAddonEssentialsToAllStorybooks(tree) {
    const projectsThatFailedTOAddAddonEssentials = [];
    (0, executor_options_utils_1.forEachExecutorOptions)(tree, '@nrwl/storybook:build', (options, projectName) => {
        const failedToAddAddon = addAddon(tree, options, projectName);
        if (failedToAddAddon) {
            projectsThatFailedTOAddAddonEssentials.push(failedToAddAddon);
        }
    });
    (0, executor_options_utils_1.forEachExecutorOptions)(tree, '@storybook/angular:build-storybook', (options, projectName) => {
        const failedToAddAddon = addAddon(tree, options, projectName);
        if (failedToAddAddon) {
            projectsThatFailedTOAddAddonEssentials.push(failedToAddAddon);
        }
    });
    return Array.from(new Set(projectsThatFailedTOAddAddonEssentials));
}
function addAddon(tree, options, projectName) {
    const storybookDir = options?.['configDir'];
    if (storybookDir) {
        const mainJsTsPath = tree.exists(`${storybookDir}/main.js`)
            ? `${storybookDir}/main.js`
            : tree.exists(`${storybookDir}/main.ts`)
                ? `${storybookDir}/main.ts`
                : undefined;
        let addedAddons = mainJsTsPath
            ? transformMainJsTs(tree, mainJsTsPath)
            : false;
        if ((storybookDir && !mainJsTsPath) || !addedAddons) {
            return projectName;
        }
    }
}
function transformMainJsTs(tree, mainJsTsPath) {
    let mainJsTs = tree.read(mainJsTsPath, 'utf-8');
    const addonsArray = tsquery_1.tsquery.query(mainJsTs, 'ArrayLiteralExpression:has(Identifier[name="addons"])')?.[0];
    if (addonsArray?.getText()?.length) {
        // If addons array does not contain @storybook/addon-essentials, add it
        if (!addonsArray.getText().includes('@storybook/addon-essentials')) {
            mainJsTs = (0, devkit_1.applyChangesToString)(mainJsTs, [
                {
                    type: devkit_1.ChangeType.Insert,
                    index: addonsArray.getStart() + 1,
                    text: `'@storybook/addon-essentials', `,
                },
            ]);
            tree.write(mainJsTsPath, mainJsTs);
            return true;
        }
        return false;
    }
    else {
        // We will add the addons array after the stories array
        // If I have a stories array, that's where my addons need to go
        // And there's no config without stories
        const storiesArray = tsquery_1.tsquery.query(mainJsTs, 'ArrayLiteralExpression:has(Identifier[name="stories"])')?.[0];
        if (storiesArray?.getText()?.length) {
            mainJsTs = (0, devkit_1.applyChangesToString)(mainJsTs, [
                {
                    type: devkit_1.ChangeType.Insert,
                    index: storiesArray.getEnd(),
                    text: `, addons: ['@storybook/addon-essentials']`,
                },
            ]);
            tree.write(mainJsTsPath, mainJsTs);
            return true;
        }
        else {
            /**
             * main.js has potentially a different structure
             * sort of like this:
             * rootMain.addons.push(' ...)
             * rootMain.stories.push(' ...)
             * Like in older versions of Nx
             */
            const { rootMainVariableName, importExpression } = getRootMainVariableName(mainJsTs);
            // If there is a PropertyAccessExpression with the text rootMain.addons
            // then check if it has addon-essentials, if not add it
            const addonsPropertyAccessExpression = tsquery_1.tsquery.query(mainJsTs, `PropertyAccessExpression:has([expression.name="${rootMainVariableName}"]):has([name="addons"])`)?.[0];
            if (rootMainVariableName && importExpression) {
                if (addonsPropertyAccessExpression?.getText() ===
                    `${rootMainVariableName}.addons.push`) {
                    const parentCallExpression = addonsPropertyAccessExpression.parent;
                    // see if parentCallExpression contains a StringLiteral with the text '@storybook/addon-essentials'
                    const hasAddonEssentials = !!tsquery_1.tsquery
                        .query(parentCallExpression, `StringLiteral:has([text="@storybook/addon-essentials"])`)?.[0]
                        ?.getText();
                    if (!hasAddonEssentials) {
                        mainJsTs = (0, devkit_1.applyChangesToString)(mainJsTs, [
                            {
                                type: devkit_1.ChangeType.Insert,
                                index: addonsPropertyAccessExpression.getEnd() + 1,
                                text: `'@storybook/addon-essentials', `,
                            },
                        ]);
                        tree.write(mainJsTsPath, mainJsTs);
                        return true;
                    }
                }
                else {
                    mainJsTs = (0, devkit_1.applyChangesToString)(mainJsTs, [
                        {
                            type: devkit_1.ChangeType.Insert,
                            index: importExpression.getEnd() + 1,
                            text: `${rootMainVariableName}.addons.push('@storybook/addon-essentials');`,
                        },
                    ]);
                    tree.write(mainJsTsPath, mainJsTs);
                    return true;
                }
            }
        }
        return false;
    }
}
function removeStoriesArrayFromRootIfEmpty(tree, rootMainJsTsPath) {
    if (rootMainJsTsPath) {
        let rootMainJsTs = tree.read(rootMainJsTsPath, 'utf-8');
        const fullStoriesNode = tsquery_1.tsquery.query(rootMainJsTs, 'PropertyAssignment:has([name="stories"])')?.[0];
        if (!fullStoriesNode) {
            return false;
        }
        const stringLiterals = tsquery_1.tsquery.query(fullStoriesNode, 'StringLiteral');
        if (stringLiterals?.length === 0) {
            rootMainJsTs = (0, devkit_1.applyChangesToString)(rootMainJsTs, [
                {
                    type: devkit_1.ChangeType.Delete,
                    start: fullStoriesNode.getStart(),
                    length: rootMainJsTs[fullStoriesNode.getEnd()] === ','
                        ? fullStoriesNode.getText().length + 1
                        : fullStoriesNode.getText().length,
                },
            ]);
            tree.write(rootMainJsTsPath, rootMainJsTs);
            return true;
        }
    }
}
function getRootMainVariableName(mainJsTs) {
    const requireVariableStatement = tsquery_1.tsquery.query(mainJsTs, `VariableStatement:has(CallExpression:has(Identifier[name="require"]))`);
    let rootMainVariableName;
    let importExpression;
    if (requireVariableStatement.length) {
        importExpression = requireVariableStatement.find((statement) => {
            const requireCallExpression = tsquery_1.tsquery.query(statement, 'CallExpression:has(Identifier[name="require"])');
            return requireCallExpression?.[0]?.getText()?.includes('.storybook/main');
        });
        if (importExpression) {
            rootMainVariableName = tsquery_1.tsquery
                .query(importExpression, 'Identifier')?.[0]
                ?.getText();
        }
    }
    else {
        const importDeclarations = tsquery_1.tsquery.query(mainJsTs, 'ImportDeclaration');
        importExpression = importDeclarations.find((statement) => {
            const stringLiteral = tsquery_1.tsquery.query(statement, 'StringLiteral');
            return stringLiteral?.[0]?.getText()?.includes('.storybook/main');
        });
        if (importExpression) {
            rootMainVariableName = tsquery_1.tsquery
                .query(importExpression, 'ImportSpecifier')?.[0]
                ?.getText();
        }
    }
    return {
        rootMainVariableName,
        importExpression,
    };
}
exports.getRootMainVariableName = getRootMainVariableName;
