// eslint-plugin-ngrx/dist/rules/effects/prefer-action-creator-in-of-type.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageId = void 0;
const path_1 = __importDefault(require("path"));
const rule_creator_1 = require("../../rule-creator");
exports.messageId = 'preferActionCreatorInOfType';
exports.default = (0, rule_creator_1.createRule)({
    name: path_1.default.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'effects',
        docs: {
            description: 'Using `action creator` in `ofType` is preferred over `string`.',
            recommended: 'warn',
        },
        schema: [],
        messages: {
            [exports.messageId]: 'Using `string` is forbidden. Use `action creator` instead.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        return {
            [`CallExpression[callee.name='ofType'] Literal`](node) {
                context.report({
                    node,
                    messageId: exports.messageId,
                });
            },
        };
    },
});
