// eslint-plugin-ngrx/dist/rules/store/good-action-hygiene.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageId = void 0;
const path_1 = __importDefault(require("path"));
const rule_creator_1 = require("../../rule-creator");
const utils_1 = require("../../utils");
exports.messageId = 'goodActionHygiene';
exports.default = (0, rule_creator_1.createRule)({
    name: path_1.default.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'store',
        docs: {
            description: 'Ensures the use of good action hygiene.',
            recommended: 'warn',
        },
        schema: [],
        messages: {
            [exports.messageId]: 'Action type `{{ actionType }}` does not follow the good action hygiene practice, use "[Source] {{ actionType }}" to define action types.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        const sourceEventPattern = /[\[].*[\]]\s.*/;
        return {
            [utils_1.actionCreatorWithLiteral]({ arguments: [node], }) {
                const { value: actionType } = node;
                if (sourceEventPattern.test(actionType)) {
                    return;
                }
                context.report({
                    node,
                    messageId: exports.messageId,
                    data: {
                        actionType,
                    },
                });
            },
        };
    },
});
