// eslint-plugin-ngrx/dist/rules/store/prefer-selector-in-select.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageId = void 0;
const path_1 = __importDefault(require("path"));
const rule_creator_1 = require("../../rule-creator");
const utils_1 = require("../../utils");
exports.messageId = 'preferSelectorInSelect';
exports.default = (0, rule_creator_1.createRule)({
    name: path_1.default.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'store',
        docs: {
            description: 'Using a selector in the `select` is preferred over `string` or `props drilling`.',
            recommended: 'warn',
        },
        schema: [],
        messages: {
            [exports.messageId]: 'Using `string` or `props drilling` is forbidden. Use a selector instead.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        const { identifiers = [] } = (0, utils_1.getNgRxStores)(context);
        const storeNames = identifiers.length > 0 ? (0, utils_1.asPattern)(identifiers) : null;
        if (!storeNames) {
            return {};
        }
        return {
            [`${(0, utils_1.pipeableSelect)(storeNames)}, ${(0, utils_1.selectExpression)(storeNames)}`](node) {
                for (const argument of node.arguments) {
                    if (!(0, utils_1.isLiteral)(argument) &&
                        !(0, utils_1.isArrowFunctionExpression)(argument) &&
                        !(0, utils_1.isFunctionExpression)(argument)) {
                        break;
                    }
                    context.report({
                        node: argument,
                        messageId: exports.messageId,
                    });
                }
            },
        };
    },
});
