// eslint-plugin-ngrx/dist/rules/store/prefer-action-creator-in-dispatch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageId = void 0;
const path_1 = __importDefault(require("path"));
const rule_creator_1 = require("../../rule-creator");
const utils_1 = require("../../utils");
exports.messageId = 'preferActionCreatorInDispatch';
exports.default = (0, rule_creator_1.createRule)({
    name: path_1.default.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'store',
        docs: {
            description: 'Using `action creator` in `dispatch` is preferred over `object` or old `Action`.',
            recommended: 'warn',
        },
        schema: [],
        messages: {
            [exports.messageId]: 'Using `object` or old `Action` is forbidden. Use `action creator` instead.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        const { identifiers = [] } = (0, utils_1.getNgRxStores)(context);
        const storeNames = identifiers.length > 0 ? (0, utils_1.asPattern)(identifiers) : null;
        if (!storeNames) {
            return {};
        }
        return {
            [`${(0, utils_1.dispatchExpression)(storeNames)} :matches(NewExpression, :not(NewExpression) > ObjectExpression)`](node) {
                const nearestUpperCallExpression = (0, utils_1.getNearestUpperNodeFrom)(node, utils_1.isCallExpression);
                const isStoreDispatchImmediateParent = nearestUpperCallExpression !== undefined &&
                    (0, utils_1.isCallExpressionWith)(nearestUpperCallExpression, storeNames, 'dispatch');
                if (!isStoreDispatchImmediateParent) {
                    return;
                }
                context.report({
                    node,
                    messageId: exports.messageId,
                });
            },
        };
    },
});
