// eslint-plugin-ngrx/dist/rules/store/prefer-action-creator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageId = void 0;
const path_1 = __importDefault(require("path"));
const rule_creator_1 = require("../../rule-creator");
exports.messageId = 'preferActionCreator';
exports.default = (0, rule_creator_1.createRule)({
    name: path_1.default.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'store',
        docs: {
            description: 'Using `action creator` is preferred over `Action class`.',
            recommended: 'warn',
        },
        schema: [],
        messages: {
            [exports.messageId]: 'Using `Action class` is forbidden. Use `action creator` instead.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        return {
            [`ClassDeclaration:has(TSClassImplements:matches([expression.name='Action'], [expression.property.name='Action'])):has(PropertyDefinition[key.name='type'])`](node) {
                context.report({
                    node,
                    messageId: exports.messageId,
                });
            },
        };
    },
});
