// @telenet/wink-core/dist/collection/components/patterns/title-text/title-text.stories.js
import { html } from "lit-html";
import * as TitleStory from "../../title/title.stories";
import * as ParagraphStory from "../../paragraph/paragraph.stories";
import { TITLE_SIZES } from "../../title/models/title.types";
import { PARAGRAPH_APPEARANCES, PARAGRAPH_SIZES } from "../../paragraph/models/paragraph.types";
import { BADGES } from "../../../models/badges.types";
export default {
    title: 'PATTERNS/TitleText',
    argTypes: {
        ...TitleStory.default.argTypes,
        ...ParagraphStory.default.argTypes,
    },
    args: {
        htmlTag: TITLE_SIZES.h1,
        text: 'The network of the future is already here',
        spacingY: 'default',
        appearance: PARAGRAPH_APPEARANCES.body,
        size: PARAGRAPH_SIZES.regular,
        slot: "No lack of variation and challenges with a job in our shops or contact centres. Every day is different, and that's how you like it. You don't simply help customers; you happily roll out the red carpet for them. While you are there for our customers, we are there for each other. With ‘a ten for social, friendly working environment’, a meaningful job and the possibility to work from home.",
    },
    parameters: {
        badges: [BADGES.inProgress],
        controls: {
            sort: 'none',
            exclude: ['spacingY'],
        },
    },
    render: (args) => html `
    <wink-title html-tag="${args.htmlTag}" text="${args.text}"></wink-title>
    <wink-paragraph appearance="${args.appearance}" size="${args.size}">${args.slot}</wink-paragraph>
  `,
};
export const Default = {};
