// udl/dist/models/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var analytics_call_fire_strategy_enum_1 = require("../enums/analytics-call-fire-strategy.enum");
var utility_1 = require("../util/utility");
var Base = /** @class */ (function () {
    function Base(directCallLabel) {
        var _this = this;
        this.attributes = {};
        this.forceResolvePromiseAfterDelay = 0; // circuitBreakerTimeoutInMS
        this.strategy = analytics_call_fire_strategy_enum_1.AnalyticsCallFireStrategyEnum.FIRE_ON_SAME_PAGE;
        this.setIdentifier = function (identifier) {
            _this.identifier = identifier;
        };
        this.getIdentifier = function () {
            return _this.identifier;
        };
        this.addAttribute = function (name, value) {
            _this.attributes[name] = value;
        };
        this.getAttribute = function (name) {
            if (_this.attributes) {
                return _this.attributes[name];
            }
        };
        this.setAttributes = function (attributes) {
            _this.attributes = attributes;
        };
        this.getAttributes = function () {
            return _this.attributes;
        };
        this.setDirectCallLabel = function (directCallLabel) {
            _this.directCallLabel = directCallLabel;
        };
        this.getDirectCallLabel = function () {
            return _this.directCallLabel;
        };
        this.setForceResolvePromiseAfterDelay = function (timeoutInMS) {
            _this.forceResolvePromiseAfterDelay = timeoutInMS;
        };
        this.getForceResolvePromiseAfterDelay = function () {
            return _this.forceResolvePromiseAfterDelay;
        };
        this.setStrategy = function (strategy) {
            _this.strategy = strategy;
        };
        this.getStrategy = function () {
            return _this.strategy;
        };
        this.directCallLabel = directCallLabel;
        this.identifier = utility_1.Utility.getUUID();
    }
    return Base;
}());
exports.Base = Base;
