// udl/dist/models/category.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Category = /** @class */ (function () {
    function Category(category) {
        var _this = this;
        this.setPrimaryCategory = function (primaryCategory) {
            _this.primaryCategory = primaryCategory;
        };
        this.getPrimaryCategory = function () {
            return _this.primaryCategory;
        };
        this.setSubCategory1 = function (subCategory1) {
            _this.subCategory1 = subCategory1;
        };
        this.getSubCategory1 = function () {
            return _this.subCategory1;
        };
        Object.assign(this, category);
    }
    return Category;
}());
exports.Category = Category;
var CategoryBuilder = /** @class */ (function () {
    function CategoryBuilder() {
        this.category = new Category();
    }
    CategoryBuilder.prototype.withPrimaryCategory = function (category) {
        this.category.setPrimaryCategory(category);
        return this;
    };
    CategoryBuilder.prototype.withSubCategoryOne = function (category) {
        this.category.setSubCategory1(category);
        return this;
    };
    CategoryBuilder.prototype.build = function () {
        return this.category;
    };
    return CategoryBuilder;
}());
exports.CategoryBuilder = CategoryBuilder;
