// udl/dist/models/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var componentInfo_1 = require("./componentInfo");
var category_1 = require("./category");
var Component = /** @class */ (function () {
    function Component(component) {
        var _this = this;
        this.getAttribute = function (name) {
            if (_this.attributes) {
                return _this.attributes[name];
            }
        };
        this.addAttribute = function (name, value) {
            if (!_this.attributes) {
                _this.attributes = {};
            }
            _this.attributes[name] = value;
        };
        this.getCategory = function () {
            return _this.category;
        };
        Object.assign(this, component);
        this.componentInfo = component ? new componentInfo_1.ComponentInfo(component.componentInfo) : new componentInfo_1.ComponentInfo();
        this.category = component ? new category_1.Category(component.category) : new category_1.Category();
    }
    Component.prototype.getComponentInfo = function () {
        return this.componentInfo;
    };
    ;
    return Component;
}());
exports.Component = Component;
var ComponentBuilder = /** @class */ (function () {
    function ComponentBuilder(component) {
        this.component = new Component(component);
    }
    ComponentBuilder.prototype.withAttribute = function (key, value) {
        this.component.addAttribute(key, value);
        return this;
    };
    ComponentBuilder.prototype.build = function () {
        return this.component;
    };
    return ComponentBuilder;
}());
exports.ComponentBuilder = ComponentBuilder;
