// udl/dist/models/item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var product_1 = require("./product");
var price_1 = require("./price");
var Item = /** @class */ (function () {
    function Item(itemJson) {
        var _this = this;
        this.setProductInfo = function (productInfo) {
            _this.productInfo = productInfo;
        };
        this.getProductInfo = function () {
            return _this.productInfo;
        };
        this.setPrice = function (price) {
            _this.price = price;
        };
        this.getPrice = function () {
            return _this.price;
        };
        this.setQuantity = function (quantity) {
            _this.quantity = quantity;
        };
        this.getQuantity = function () {
            return _this.quantity;
        };
        if (itemJson) {
            Object.assign(this, itemJson);
            this.productInfo = itemJson.productInfo ? new product_1.Product(itemJson.productInfo) : undefined;
            this.price = itemJson.price ? new price_1.Price(itemJson.price) : undefined;
        }
    }
    return Item;
}());
exports.Item = Item;
var ItemBuilder = /** @class */ (function () {
    function ItemBuilder() {
        this.item = new Item();
    }
    ItemBuilder.prototype.withProductInfo = function (productInfo) {
        this.item.setProductInfo(productInfo);
        return this;
    };
    ItemBuilder.prototype.withPrice = function (price) {
        this.item.setPrice(price);
        return this;
    };
    ItemBuilder.prototype.withQuantity = function (quantity) {
        this.item.setQuantity(quantity);
        return this;
    };
    ItemBuilder.prototype.build = function () {
        return this.item;
    };
    return ItemBuilder;
}());
exports.ItemBuilder = ItemBuilder;
