// udl/dist/models/pageInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PageInfo = /** @class */ (function () {
    function PageInfo(pageInfo) {
        var _this = this;
        this.getPageID = function () {
            return _this.pageID;
        };
        this.setPageID = function (pageId) {
            _this.pageID = pageId;
        };
        this.setPageName = function (pageName) {
            _this.pageName = pageName;
        };
        this.getPageName = function () {
            return _this.pageName;
        };
        this.setPageTitle = function (pageTitle) {
            _this.pageTitle = pageTitle;
        };
        this.getPageTitle = function () {
            return _this.pageTitle;
        };
        this.setPageType = function (pageType) {
            _this.pageType = pageType;
        };
        this.getPageType = function () {
            return _this.pageType;
        };
        this.setReferringURL = function (referringURL) {
            _this.referringURL = referringURL;
        };
        this.getReferringURL = function () {
            return _this.referringURL;
        };
        this.setDestinationURL = function (destinationURL) {
            _this.destinationURL = destinationURL;
        };
        this.getDestinationURL = function () {
            return _this.destinationURL;
        };
        this.setSysEnv = function (sysEnv) {
            _this.sysEnv = sysEnv;
        };
        this.getSysEnv = function () {
            return _this.sysEnv;
        };
        this.setLanguage = function (language) {
            _this.language = language;
        };
        this.getLanguage = function () {
            return _this.language;
        };
        this.setEffectiveDate = function (date) {
            _this.effectiveDate = date;
        };
        this.getEffectiveDate = function () {
            return _this.effectiveDate;
        };
        this.setExpiryDate = function (date) {
            _this.expiryDate = date;
        };
        this.getExpiryDate = function () {
            return _this.expiryDate;
        };
        this.setIssueDate = function (date) {
            _this.issueDate = date;
        };
        this.getIssueDate = function () {
            return _this.issueDate;
        };
        this.setPreviousPageName = function (previousPageName) {
            _this.previousPageName = previousPageName;
        };
        this.getPreviousPageName = function () {
            return _this.previousPageName;
        };
        this.setStepId = function (stepId) {
            _this.stepId = stepId;
        };
        this.getStepId = function () {
            return _this.stepId;
        };
        Object.assign(this, pageInfo);
    }
    return PageInfo;
}());
exports.PageInfo = PageInfo;
var PageInfoBuilder = /** @class */ (function () {
    function PageInfoBuilder(pageInfo) {
        this.pageInfo = new PageInfo(pageInfo);
    }
    PageInfoBuilder.prototype.withPageId = function (pageId) {
        this.pageInfo.setPageID(pageId);
        return this;
    };
    PageInfoBuilder.prototype.withPageName = function (pageName) {
        this.pageInfo.setPageName(pageName);
        return this;
    };
    PageInfoBuilder.prototype.withPageTitle = function (pageTitle) {
        this.pageInfo.setPageTitle(pageTitle);
        return this;
    };
    PageInfoBuilder.prototype.withPageType = function (pageType) {
        this.pageInfo.setPageType(pageType);
        return this;
    };
    PageInfoBuilder.prototype.withReferringURL = function (referringURL) {
        this.pageInfo.setReferringURL(referringURL);
        return this;
    };
    PageInfoBuilder.prototype.withDestionationURL = function (destinationURL) {
        this.pageInfo.setDestinationURL(destinationURL);
        return this;
    };
    PageInfoBuilder.prototype.withSysEnv = function (sysEnv) {
        this.pageInfo.setSysEnv(sysEnv);
        return this;
    };
    PageInfoBuilder.prototype.withLanguage = function (language) {
        this.pageInfo.setLanguage(language);
        return this;
    };
    PageInfoBuilder.prototype.withEffectiveDate = function (date) {
        this.pageInfo.setEffectiveDate(date);
        return this;
    };
    PageInfoBuilder.prototype.withExpiryDate = function (date) {
        this.pageInfo.setExpiryDate(date);
        return this;
    };
    PageInfoBuilder.prototype.withIssueDate = function (date) {
        this.pageInfo.setIssueDate(date);
        return this;
    };
    PageInfoBuilder.prototype.withPreviousPageName = function (previousPageName) {
        this.pageInfo.setPreviousPageName(previousPageName);
        return this;
    };
    PageInfoBuilder.prototype.withStepId = function (stepId) {
        this.pageInfo.setStepId(stepId);
        return this;
    };
    PageInfoBuilder.prototype.build = function () {
        return this.pageInfo;
    };
    return PageInfoBuilder;
}());
exports.PageInfoBuilder = PageInfoBuilder;
