// udl/dist/models/page.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var pageInfo_1 = require("./pageInfo");
var base_1 = require("./base");
var direct_call_label_enum_1 = require("../enums/direct-call-label.enum");
var Page = /** @class */ (function (_super) {
    __extends(Page, _super);
    function Page(page) {
        var _this = _super.call(this, direct_call_label_enum_1.DirectCallLabelEnum.PAGE_LOADED) || this;
        _this.getPageInfo = function () {
            return _this.pageInfo;
        };
        Object.assign(_this, page);
        _this.setIdentifier(Page.PAGE_IDENTIFIER);
        _this.pageInfo = page ? new pageInfo_1.PageInfo(page.pageInfo) : new pageInfo_1.PageInfo();
        return _this;
    }
    Page.PAGE_IDENTIFIER = 'page-identifier';
    return Page;
}(base_1.Base));
exports.Page = Page;
var PageBuilder = /** @class */ (function () {
    function PageBuilder(page) {
        this.page = new Page(page);
    }
    PageBuilder.prototype.withAttribute = function (key, value) {
        this.page.addAttribute(key, value);
        return this;
    };
    PageBuilder.prototype.withDirectCallLabel = function (directCallLabel) {
        this.page.setDirectCallLabel(directCallLabel);
        return this;
    };
    PageBuilder.prototype.build = function () {
        return this.page;
    };
    return PageBuilder;
}());
exports.PageBuilder = PageBuilder;
