// udl/dist/models/productHolding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ProductHolding = /** @class */ (function () {
    function ProductHolding() {
        var _this = this;
        this.setProductID = function (productID) {
            _this.productID = productID;
        };
        this.getProductID = function () {
            return _this.productID;
        };
        this.setTvBoxType = function (tvBoxType) {
            _this.tvBoxType = tvBoxType;
        };
        this.getTvBoxType = function () {
            return _this.tvBoxType;
        };
        this.setServices = function (services) {
            _this.services = services;
        };
        this.getServices = function () {
            return _this.services;
        };
        this.setLoggedInUserProductCategory = function (loggedInUserProductCategory) {
            _this.loggedInUserProductCategory = loggedInUserProductCategory;
        };
        this.getLoggedInUserProductCategory = function () {
            return _this.loggedInUserProductCategory;
        };
    }
    return ProductHolding;
}());
exports.ProductHolding = ProductHolding;
var ProductHoldingBuilder = /** @class */ (function () {
    function ProductHoldingBuilder() {
        this.productHolding = new ProductHolding();
    }
    ProductHoldingBuilder.prototype.withProductId = function (productId) {
        this.productHolding.setProductID(productId);
        return this;
    };
    ProductHoldingBuilder.prototype.withTvBoxType = function (tvBoxType) {
        this.productHolding.setTvBoxType(tvBoxType);
        return this;
    };
    ProductHoldingBuilder.prototype.withServices = function (services) {
        this.productHolding.setServices(services);
        return this;
    };
    ProductHoldingBuilder.prototype.withLoggedInUserProductCategory = function (loggedInUserProductCategory) {
        this.productHolding.setLoggedInUserProductCategory(loggedInUserProductCategory);
        return this;
    };
    ProductHoldingBuilder.prototype.build = function () {
        return this.productHolding;
    };
    return ProductHoldingBuilder;
}());
exports.ProductHoldingBuilder = ProductHoldingBuilder;
