// udl/dist/models/product.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Product = /** @class */ (function () {
    function Product(productJson) {
        var _this = this;
        this.productType = 'NotAvailable';
        this.setProductID = function (id) {
            _this.productID = id;
        };
        this.getProductID = function () {
            return _this.productID;
        };
        this.setProductName = function (name) {
            _this.productName = name;
        };
        this.getProductName = function () {
            return _this.productName;
        };
        this.setProductType = function (type) {
            _this.productType = type;
        };
        this.getProductType = function () {
            return _this.productType;
        };
        this.setProductStatus = function (productStatus) {
            _this.productStatus = productStatus;
        };
        this.getProductStatus = function () {
            return _this.productStatus;
        };
        this.setProductStock = function (productStock) {
            _this.productStock = productStock;
        };
        this.getProductStock = function () {
            return _this.productStock;
        };
        Object.assign(this, productJson);
    }
    return Product;
}());
exports.Product = Product;
var ProductBuilder = /** @class */ (function () {
    function ProductBuilder() {
        this.product = new Product();
    }
    ProductBuilder.prototype.withProductId = function (id) {
        this.product.setProductID(id);
        return this;
    };
    ProductBuilder.prototype.withProductName = function (name) {
        this.product.setProductName(name);
        return this;
    };
    ProductBuilder.prototype.withProductType = function (type) {
        this.product.setProductType(type);
        return this;
    };
    ProductBuilder.prototype.withProductStatus = function (status) {
        this.product.setProductStatus(status);
        return this;
    };
    ProductBuilder.prototype.withProductStock = function (stock) {
        this.product.setProductStock(stock);
        return this;
    };
    ProductBuilder.prototype.build = function () {
        return this.product;
    };
    return ProductBuilder;
}());
exports.ProductBuilder = ProductBuilder;
