// udl/dist/models/profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Profile = /** @class */ (function () {
    function Profile() {
        var _this = this;
        this.setLoginState = function (loginstate) {
            _this.loginstate = loginstate;
        };
        this.getLoginState = function () {
            return _this.loginstate;
        };
        this.setLoginScope = function (loginscope) {
            _this.loginscope = loginscope;
        };
        this.getLoginScope = function () {
            return _this.loginscope;
        };
        this.setImpersonation = function (impersonation) {
            _this.impersonation = impersonation;
        };
        this.getImpersonation = function () {
            return _this.impersonation;
        };
        this.setTechnicalID = function (technicalID) {
            _this.technicalID = technicalID;
        };
        this.getTechnicalID = function () {
            return _this.technicalID;
        };
        this.setPostalCode = function (postalcode) {
            _this.postalcode = postalcode;
        };
        this.getPostalCode = function () {
            return _this.postalcode;
        };
        this.setInstallationPostalCode = function (installationPostalCode) {
            _this.installationPostalCode = installationPostalCode;
        };
        this.getInstallationPostalCode = function () {
            return _this.installationPostalCode;
        };
        this.setBillingPostalCode = function (billingPostalCode) {
            _this.billingPostalCode = billingPostalCode;
        };
        this.getBillingPostalCode = function () {
            return _this.billingPostalCode;
        };
        this.setCustomerCategory = function (customerCategory) {
            _this.customerCategory = customerCategory;
        };
        this.getCustomerCategory = function () {
            return _this.customerCategory;
        };
        this.setLoginType = function (logintype) {
            _this.logintype = logintype;
        };
        this.getLoginType = function () {
            return _this.logintype;
        };
        this.setCustomerType = function (customerType) {
            _this.customerType = customerType;
        };
        this.getCustomerType = function () {
            return _this.customerType;
        };
    }
    return Profile;
}());
exports.Profile = Profile;
var ProfileBuilder = /** @class */ (function () {
    function ProfileBuilder() {
        this.profile = new Profile();
    }
    ProfileBuilder.prototype.withLoginState = function (loginstate) {
        this.profile.setLoginState(loginstate);
        return this;
    };
    ProfileBuilder.prototype.withLoginScope = function (loginscope) {
        this.profile.setLoginScope(loginscope);
        return this;
    };
    ProfileBuilder.prototype.withImpersonation = function (impersonation) {
        this.profile.setImpersonation(impersonation);
        return this;
    };
    ProfileBuilder.prototype.withTechnicalID = function (technicalID) {
        this.profile.setTechnicalID(technicalID);
        return this;
    };
    ;
    ProfileBuilder.prototype.withPostalCode = function (postalcode) {
        this.profile.setPostalCode(postalcode);
        return this;
    };
    ;
    ProfileBuilder.prototype.withInstallationPostalCode = function (installationPostalCode) {
        this.profile.setInstallationPostalCode(installationPostalCode);
        return this;
    };
    ;
    ProfileBuilder.prototype.withBillingPostalCode = function (billingPostalCode) {
        this.profile.setBillingPostalCode(billingPostalCode);
        return this;
    };
    ;
    ProfileBuilder.prototype.withCustomerCategory = function (customerCategory) {
        this.profile.setCustomerCategory(customerCategory);
        return this;
    };
    ;
    ProfileBuilder.prototype.withLoginType = function (logintype) {
        this.profile.setLoginType(logintype);
        return this;
    };
    ;
    ProfileBuilder.prototype.withCustomerType = function (customerType) {
        this.profile.setCustomerType(customerType);
        return this;
    };
    ;
    ProfileBuilder.prototype.build = function () {
        return this.profile;
    };
    return ProfileBuilder;
}());
exports.ProfileBuilder = ProfileBuilder;
