// udl/dist/models/user.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = require("./base");
var direct_call_label_enum_1 = require("../enums/direct-call-label.enum");
var User = /** @class */ (function (_super) {
    __extends(User, _super);
    function User() {
        var _this = _super.call(this, direct_call_label_enum_1.DirectCallLabelEnum.USER_DATA_LOADED) || this;
        _this.setSegment = function (segment) {
            _this.segment = segment;
        };
        _this.getSegment = function () {
            return _this.segment;
        };
        _this.setProfile = function (profile) {
            _this.profile = profile;
        };
        _this.getProfile = function () {
            return _this.profile;
        };
        _this.setProductHolding = function (productHolding) {
            _this.productHolding = productHolding;
        };
        _this.getProductHolding = function () {
            return _this.productHolding;
        };
        _this.setBilling = function (billing) {
            _this.billing = billing;
        };
        _this.getBilling = function () {
            return _this.billing;
        };
        _this.setPega = function (pega) {
            _this.pega = pega;
        };
        _this.getPega = function () {
            return _this.pega;
        };
        return _this;
    }
    return User;
}(base_1.Base));
exports.User = User;
var UserBuilder = /** @class */ (function () {
    function UserBuilder() {
        this.user = new User();
    }
    UserBuilder.prototype.withSegment = function (segment) {
        this.user.setSegment(segment);
        return this;
    };
    UserBuilder.prototype.withProfile = function (profile) {
        this.user.setProfile(profile);
        return this;
    };
    UserBuilder.prototype.withProductHolding = function (productHolding) {
        this.user.setProductHolding(productHolding);
        return this;
    };
    UserBuilder.prototype.withDirectCallLabel = function (directCallLabel) {
        this.user.setDirectCallLabel(directCallLabel);
        return this;
    };
    UserBuilder.prototype.withBilling = function (billing) {
        this.user.setBilling(billing);
        return this;
    };
    UserBuilder.prototype.withPega = function (pega) {
        this.user.setPega(pega);
        return this;
    };
    UserBuilder.prototype.build = function () {
        return this.user;
    };
    return UserBuilder;
}());
exports.UserBuilder = UserBuilder;
